task 1: annotate

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* #61188
* #61150
* #61093
* #61012
* #60805
* #60622
* **#60621**

Differential Revision: [D29493619](https://our.internmc.facebook.com/intern/diff/D29493619)

migeed-z committed:
task 1: annotate

[ghstack-poisoned]

jamesr66a commented on test/test_fx.py:
```diff
@@ -35,6 +35,7 @@
 from fx.test_subgraph_rewriter import TestSubgraphRewriter  # noqa: F401
 from fx.test_dce_pass import TestDCE  # noqa: F401
 from fx.test_fx_const_fold import TestConstFold  # noqa: F401
+from fx.test_gradual_type import AnnotationsTest  #  noqa: F401
```
As discussed offline, because python below 3.7 does not have `__class_getitem__`, let's disable this feature for unsupported python versions. Let's gate this like so:

```
if sys.version_info >= (3, 7):
  from fx.test_gradual_type import AnnotationsTest  #  noqa: F401
```

migeed-z committed:
Update on "task 1: annotate"




[ghstack-poisoned]

migeed-z committed:
Update on "task 1: annotate"




[ghstack-poisoned]

migeed-z committed:
Update on "task 1: annotate"




[ghstack-poisoned]

migeed-z committed:
Update on "task 1: annotate"




[ghstack-poisoned]

jamesr66a commented on torch/fx/annotate.py:
```diff
@@ -0,0 +1,13 @@
+    # or fx.Proxy (tracing)
+    if isinstance(val, Proxy):
+        if val.node.type:
+            raise RuntimeError("type already exists")
```
A good error message will direct the user on what they can do to fix the issue. I recommend something like

```
raise RuntimeError(f"Tried to annotate a value that already had a type on it! Existing type is {val.node.type} "
                   f"and new type is {type}. This could happen if you tried to annotate a function parameter "
                   f"value (in which case you should use the type slot on the function signature) or you called "
                   f"annotate on the same value twice")
```

jamesr66a commented on torch/fx/tensor_type.py:
```diff
@@ -0,0 +1,62 @@
+        self.__args__ = dim
+
+    def __repr__(self):
+        return f'Tensor_Type[{self.__args__}]'
```
nit: for consistency, printout should be `TensorType`

jamesr66a commented on torch/fx/tensor_type.py:
```diff
@@ -0,0 +1,62 @@
+    _DynType defines a type which stands for the absence of type information.
+    """
+    def __init__(self):
+        self.__name__ = '_DynType'
```
Is this for printing it out or something else?

I think the more canonical way of defining the printouts is to define a `__repr__` method

migeed-z replied:
if I remove the name attribute the symbolic execution complains 
`AttributeError: '_DynType' object has no attribute '__name__'`

jamesr66a replied:
Can you post the full error? (including stack trace)

That probably indicates that the prettyprint process for types is expecting that attribute

migeed-z replied:
Error
Traceback (most recent call last):
  File "/Users/migeedz/pytorch/test/fx/test_gradual_type.py", line 38, in test_annotate
    symbolic_traced : torch.fx.GraphModule = symbolic_trace(module)
  File "/Users/migeedz/pytorch/torch/fx/_symbolic_trace.py", line 863, in symbolic_trace
    return GraphModule(tracer.root, graph, name)
  File "/Users/migeedz/pytorch/torch/fx/graph_module.py", line 234, in __init__
    self.graph = graph
  File "/Users/migeedz/pytorch/torch/nn/modules/module.py", line 1179, in __setattr__
    object.__setattr__(self, name, value)
  File "/Users/migeedz/pytorch/torch/fx/graph_module.py", line 259, in graph
    self.recompile()
  File "/Users/migeedz/pytorch/torch/fx/graph_module.py", line 467, in recompile
    python_code = self._graph.python_code(root_module='self')
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 803, in python_code
    return self._python_code(root_module, namespace)
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 955, in _python_code
    emit_node(node)
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 897, in emit_node
    maybe_type_annotation = '' if node.type is None else f' : {type_repr(node.type)}'
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 854, in type_repr
    args = [type_repr(arg) for arg in o.__args__]
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 854, in <listcomp>
    args = [type_repr(arg) for arg in o.__args__]
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 859, in type_repr
    return add_global(typename, o)
  File "/Users/migeedz/pytorch/torch/fx/graph.py", line 826, in add_global
    return _get_qualified_name(obj)
  File "/Users/migeedz/pytorch/torch/fx/node.py", line 64, in _get_qualified_name
    if getattr(builtins, func.__name__, None) is func:
AttributeError: '_DynType' object has no attribute '__name__'

jamesr66a reviewed (approved):
Looks good! Just some tiny comments and you should be good to go

migeed-z committed:
Update on "task 1: annotate"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29493619).

migeed-z committed:
Update on "task 1: annotate"


Differential Revision: [D29493619](https://our.internmc.facebook.com/intern/diff/D29493619)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29493619).
